crates/wasmparser/src/binary_reader.rs

```diff
@@ -354,20 +361,26 @@ impl<'a> BinaryReader<'a> {
     pub(crate) fn read_memory_type(&mut self) -> Result<MemoryType> {
         let pos = self.original_position();
         let flags = self.read_u8()?;
-        if (flags & !0x7) != 0 {
-            return Err(BinaryReaderError::new(
-                "invalid table resizable limits flags",
-                pos,
-            ));
+        if (flags & !0b111) != 0 {
+            return Err(BinaryReaderError::new("invalid memory limits flags", pos));
         }
-        if flags & 0x4 == 0 {
-            let limits = self.read_resizable_limits((flags & 0x1) != 0)?;
-            let shared = (flags & 0x2) != 0;
-            Ok(MemoryType::M32 { limits, shared })
-        } else {
-            let limits = self.read_resizable_limits64((flags & 0x1) != 0)?;
-            let shared = (flags & 0x2) != 0;
+
+        let b64 = flags & 0b100 != 0;
+        let shared = flags & 0b010 != 0;
+        let has_max = flags & 0b001 != 0;
+        if b64 {
+            // FIXME(WebAssembly/memory64#21) as currently specified if the
+            // `shared` flag is set we should be reading a 32-bit limits field
+            // here. That seems a bit odd to me at the time of this writing so
+            // I've taken the liberty of reading a 64-bit limits field in those
+            // situations. I suspect that this is a typo in the spec, but if not
+            // we'll need to update this to read a 32-bit limits field when the
+            // shared flag is set.
+            let limits = self.read_resizable_limits64(has_max)?;
             Ok(MemoryType::M64 { limits, shared })
+        } else {
+            let limits = self.read_resizable_limits(has_max)?;
+            Ok(MemoryType::M32 { limits, shared })
```

Comment:
It seems strange, API-wise, that we have `enum { M32, M64 }` for `MemoryType` but just have `u64`s inside `MemArg` instead of also having a 32 vs 64 `enum`.

In my experience, the `enum { M32, M64 }` pattern is quite annoying to work with. What do you think of switching `MemoryType` over into a `struct` with a `is_memory64: bool` field and always have 64-bit limits?

Not 100% convinced this is a good idea, but it would be nice to unify `MemoryType` and `MemArg`'s approach to dealing with memory64, and I'm not in love with the idea of making `MemArg` into an `enum` either.

Reply:
Nah makes sense and I agree that the `M32`/`M64` business was generally just a pain to deal with.